src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -2577,11 +2608,30 @@ static void finish_bdp_ping_locked(grpc_exec_ctx *exec_ctx, void *tp,
                                    grpc_error *error) {
   grpc_chttp2_transport *t = (grpc_chttp2_transport *)tp;
   if (GRPC_TRACER_ON(grpc_http_trace)) {
-    gpr_log(GPR_DEBUG, "%s: Complete BDP ping", t->peer_string);
+    gpr_log(GPR_DEBUG, "%s: Complete BDP ping err=%s", t->peer_string,
+            grpc_error_string(error));
+  }
+  if (error != GRPC_ERROR_NONE) {
+    GRPC_CHTTP2_UNREF_TRANSPORT(exec_ctx, t, "bdp_ping");
+    return;
   }
-  t->flow_control.bdp_estimator->CompletePing(exec_ctx);
+  grpc_millis next_ping = t->flow_control.bdp_estimator->CompletePing(exec_ctx);
+  GPR_ASSERT(!t->have_next_bdp_ping_timer);
+  t->have_next_bdp_ping_timer = true;
+  grpc_timer_init(exec_ctx, &t->next_bdp_ping_timer, next_ping,
```

Comment:
What lock are we under when the timer pops, t->combiner? PRobably basic question, I am not too familiar with our timers yet

Reply:
We initialize the closure against the combiner lock earlier:
```
  GRPC_CLOSURE_INIT(&t->next_bdp_ping_timer_expired_locked,
                    next_bdp_ping_timer_expired_locked, t,
                    grpc_combiner_scheduler(t->combiner));
```

The timer code just schedules this closure at an appropriate time